subsys/bluetooth/mesh/shell_util.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ctype.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <errno.h>
+#include "shell_util.h"
+
+static size_t whitespace_trim(char *out, size_t len, const char *str)
+{
+	if (len == 0) {
+		return 0;
+	}
+
+	const char *end;
+	size_t out_size;
+
+	while (str[0] == ' ') {
+		str++;
+	}
+
+	if (*str == 0) {
+		*out = 0;
+		return 1;
+	}
+
+	end = str + strlen(str) - 1;
+	while (end > str && (end[0] == ' ')) {
+		end--;
+	}
+	end++;
+
+	out_size = (end - str) + 1;
+
+	if (out_size > len) {
+		return 0;
+	}
+
+	memcpy(out, str, out_size - 1);
+	out[out_size - 1] = 0;
+
+	return out_size;
+}
+
+int bt_mesh_str2l(const char *str, long *val)
+{
+	char trimmed_buf[12] = { 0 };
+	size_t len = whitespace_trim(trimmed_buf, sizeof(trimmed_buf), str);
+
+	if (len < 2) {
+		return -EINVAL;
+	}
+
+	long temp_val;
+	int idx = 0;
+
+	if ((trimmed_buf[0] == '-') || (trimmed_buf[0] == '+')) {
+		idx++;
+	}
+
+	for (int i = idx; i < (len - 1); i++) {
+		if (!isdigit((int)trimmed_buf[i])) {
+			return -EINVAL;
+		}
+	}
+
+	errno = 0;
+	temp_val = strtol(trimmed_buf, NULL, 10);
+
+	if (errno == ERANGE) {
+		return -ERANGE;
+	}
+
+	*val = temp_val;
+	return 0;
```

Comment:
This can be simplified a lot. For one, `strtol()` wil skip leading whitespace completely on its own. It also convenientely returns the endptr for you if you provide it, which can be used for checking that only whitespace remains after the number. Combination of these makes this (and the unsigned version) simplify to
```c
char *endptr = NULL;

errno = 0;
*val = strtol(str, &endptr, 0);
if (errno == ERANGE) {
    return -ERANGE;
} else if (errno || endptr == NULL) {
    return -EINVAL;
}

/* Check that remaining string is whitespace.
Not sure if this is really needed now that this
is scoped to shell argument usage where
whitespace is already trimmed by the shell */
while ((char c = *endptr++)) {
    if (!isspace(c)) {
        return -EINVAL;
    }
}

return 0;
```

Reply:
If we assume there should never be trailing whitespace, the while loop can be replaced with appending `|| *endptr != 0` to the `else if` above.

Reply:
Thanks for the input 👍